### PR TITLE
Provide better diagnostic for missing meta objects

### PR DESCRIPTION
**Pull request by Neverlord**
Let's be nicer to CAF users. 🙂

It's easy to forget initializing the meta objects table. When forgetting it, CAF just crashes in some internal function that the user never heard of. We know what's wrong, so not telling the user is just bad manners. Especially towards first-time CAF users that probably think CAF is just broken.

So, to improve usability of CAF, we now print a neat error message that gives users some guidance as to what went wrong.

Before: segfault. 🤨

After: 💡
```
[FATAL] critical error (/Users/neverlord/workspace/project/caf/libcaf_core/src/detail/meta_object.cpp:68): found no meta object for type ID 200!
        This usually means that run-time type initialization is missing.
        With CAF_MAIN, make sure to pass all custom type ID blocks.
        With a custom main, call (before any other CAF function):
        - caf::core::init_global_meta_objects()
        - <module>::init_global_meta_objects() for all loaded modules
        - caf::init_global_meta_objects<T>() for all custom ID blocks
```

We only have a few places where we can deal with non-existing type IDs. Mostly while deserializing some data into C++ objects, where we have to expect that the data might be corrupted. For these use cases, we have a new function now that may still return `nullptr`.

### CODE REVIEW

**Comment by alatdneg** [comment]
Nice, that confused me no end when my serialise test code bombed out.

File: libcaf_core/src/detail/meta_object.cpp [hunk 1]
```diff
@@ -54,7 +54,26 @@ span<const meta_object> global_meta_objects() {
...
+  CAF_CRITICAL_FMT(
+    "found no meta object for type ID %d!\n"
+    "        This usually means that run-time type initialization is missing.\n"
+    "        With CAF_MAIN, make sure to pass all custom type ID blocks.\n"
+    "        With a custom main, call (before any other CAF function):\n"
+    "        - caf::core::init_global_meta_objects()\n"
+    "        - <module>::init_global_meta_objects() for all loaded modules\n"
+    "        - caf::init_global_meta_objects<T>() for all custom ID blocks",
+    static_cast<int>(id));
+  return nullptr;
```
Comment (josephnoir) [inline libcaf_core/src/detail/meta_object.cpp:69]: Is this unreachable code? Which to me would mean a compiler warning and it should not be here.

File: libcaf_core/src/type_id_list.cpp [hunk 2]
```diff
@@ -13,8 +13,8 @@ namespace caf {
 size_t type_id_list::data_size() const noexcept {
   auto result = size_t{0};
   for (auto type : *this) {
-    auto meta_obj = detail::global_meta_object(type);
-    result += meta_obj->padded_size;
+    auto meta = detail::global_meta_object(type);
+    result += meta->padded_size;
```
Comment (josephnoir) [inline libcaf_core/src/type_id_list.cpp:17]: Why? In the `message.cpp` diff it is still called `meta_obj` and in the `type_id.cpp` diff the result is called `ptr`.

File: libcaf_core/src/detail/meta_object.cpp (hunk 1 above)
Comment (dominiklohmann) [reply libcaf_core/src/detail/meta_object.cpp:69]: The function could also return a reference if it's marked `[[noreturn]]` given that `CAF_CRITICAL` always ends up calling `std::abort()`.

Comment (Neverlord) [reply libcaf_core/src/detail/meta_object.cpp]: ACK to both.

File: libcaf_core/src/type_id_list.cpp (hunk 2 above)
Comment (Neverlord) [reply libcaf_core/src/type_id_list.cpp]: Should be consistently `meta` now wherever we access a meta object.

**Review by josephnoir — APPROVED** [review]

**Pull request merged**